### PR TITLE
ebib-clean-TeX-markup: generalise + use regexps

### PR DESCRIPTION
**Pull request by Hugo-Heagren**
# ebib-clean-TeX-markup: generalise + use regexps
[This](https://github.com/joostkremers/ebib/pull/238#issuecomment-1008155045) was annoying me -- I wanted it to work for proper regexps originally, so I've fixed it so that it does. Combined with the fact that the code uses `replace-regexp-in-string`, this means that the 'replacement' can be a string or a function, which means we can generalise away the more specific cases in the original code, into the variable.

This is a draft PR, because although the thing I wanted to work, works, I have some questions:
- The functions used in `ebib-TeX-markup-replace-alist` are all lambdas. Is this alright? They aren't used anywhere else, so it seemed pointless to define them, but it isn't very pretty, and if you're new to elisp their function might be less obvious than if they were well-named symbols.
- Three out of four of the functions in there at the moment just `propertize`ations. Is it worth abstracting this pattern out (e.g. by defining a third possibility for the value, so that the cdr specifies a property (and maybe a match-string to propertize), and the actual function call is done by `ebib-clean-TeX-markup`)
- I was originally thinking about this because I came across some more symbols I needed to add locally. I found a [pretty exhaustive](https://github.com/loopspace/latexToUTF/blob/master/utf.php#L3-L2237) list of conversions. Is it worth adding some more to ebib's core, like a core list of symbols lots of people might use, especially in titles? Then again, I don't know much about emac's handling of unicode, which this would seem to rely on... just a thought.

Anyway, what I've got is pretty much ready, but the above were other related considerations which go along with it.

### CODE REVIEW

**Comment by joostkremers** [comment]
> # ebib-clean-TeX-markup: generalise + use regexps
> [This](https://github.com/joostkremers/ebib/pull/238#issuecomment-1008155045) was annoying me -- I wanted it to work for proper regexps originally, so I've fixed it so that it does.

Cool!

> This is a draft PR, because although the thing I wanted to work, works, I have some questions:
> 
> * The functions used in `ebib-TeX-markup-replace-alist` are all lambdas. Is this alright? They aren't used anywhere else, so it seemed pointless to define them, but it isn't very pretty, and if you're new to elisp their function might be less obvious than if they were well-named symbols.

Even though they really are trivial, I think I'd still prefer them as properly defined functions, mainly because it's a user-configurable option.

> * Three out of four of the functions in there at the moment just `propertize`ations. Is it worth abstracting this pattern out (e.g. by defining a third possibility for the value, so that the cdr specifies a property (and maybe a match-string to propertize), and the actual function call is done by `ebib-clean-TeX-markup`)

That would be an option. Whether it's worth it, I can't say right now. If it's just these three, then perhaps not. If the pattern is something that is more generally useful, then it might be. (Right now, though, I don't see how it would be more generally useful, TBH.)

> * I was originally thinking about this because I came across some more symbols I needed to add locally. I found a [pretty exhaustive](https://github.com/loopspace/latexToUTF/blob/master/utf.php#L3-L2237) list of conversions. Is it worth adding some more to ebib's core, like a core list of symbols lots of people might use, especially in titles? Then again, I don't know much about emac's handling of unicode, which this would seem to rely on... just a thought.

Emacs' Unicode support is quite good, so that's not an issue. The problem is with fonts. If the user doesn't have a font that supports a particular character (or has one that differs visually from their default font), then the display will be ugly and/or distorted (line height and alignment may be disrupted). So I wouldn't add too many characters by default, though a few common ones might be nice.

> Anyway, what I've got is pretty much ready, but the above were other related considerations which go along with it.

Your PR doesn't seem to deal with LaTeX commands other than \textit, \emph, \textbf and \textsc. The code it's replacing removed any LaTeX commands together with any optional arguments, leaving only the obligatory arguments. I assume adding a general regex for this case to the end of `ebib-TeX-markup-replace-alist` should be able to handle that.

File: ebib-utils.el [hunk 1]
```diff
@@ -392,22 +392,38 @@ transformation function returns something that can be displayed."
...
 
-(defcustom ebib-TeX-markup-replace-alist '(("{" . "")
-                                           ("}" . "")
-                                           ("``" . "“")
+(defcustom ebib-TeX-markup-replace-alist '(("``" . "“")
 					   ("`" . "‘")
 					   ("''" . "”")
-					   ("'" . "’"))
+					   ("'" . "’")
+					   ("\\\\textit\\(?:\\[.*\\]\\)?{\\([^}]*\\)}"
```
Comment (joostkremers) [inline ebib-utils.el:399]: The regexes here differ slightly from the original regex in `ebib-clean-TeX-markup`? Did you do that for a reason?

File: ebib-utils.el [hunk 2]
```diff
@@ -392,22 +392,38 @@ transformation function returns something that can be displayed."
...
+simply replace the match) or a function (the match will be
+replaced by the result of calling the function on the match
+string).  Earlier elements are evaluated before later ones, so if
+one string is a subpattern of another, the second must appear
+later (e.g. \"''\" is before \"'\")."
   :group 'ebib
-  :type '(alist (string :tag "String")
-		(string :tag "Replacement")))
+  :type '(alist (regexp :tag "Pattern")
+		((choice string function) :tag "Replacement")))
```
Comment (joostkremers) [inline ebib-utils.el:426]: It might be nice to use a somewhat more user-friendly display in customization buffers, e.g. a set of check boxes to turn different options on or off. But that can be a later improvement, doesn't have to be in this PR.

File: ebib-utils.el [hunk 3]
```diff
@@ -2376,25 +2392,19 @@ year can be extracted from DATE, return nil."
   (ebib-clean-TeX-markup (ebib-get-field-value field key db "" 'unbraced 'xref)))
 
 (defun ebib-clean-TeX-markup (string)
-  "Return STRING without TeX markup."
+  "Return STRING without TeX markup.
+
+Any substring matching the car of a cell in
+`ebib-TeX-markup-replace-alist' is replaced with the
+correspnoding cdr, if the cdr is a string, or with the result of
```
Comment (joostkremers) [inline ebib-utils.el:2399]: I'd put `"if the cdr is a string"` in parentheses, because the equivalent remark one line below is, as well. (Or don't put parentheses around either.)

File: ebib-utils.el [hunk 4]
```diff
@@ -2376,25 +2392,19 @@ year can be extracted from DATE, return nil."
   (ebib-clean-TeX-markup (ebib-get-field-value field key db "" 'unbraced 'xref)))
 
 (defun ebib-clean-TeX-markup (string)
-  "Return STRING without TeX markup."
+  "Return STRING without TeX markup.
+
```
Comment (joostkremers) [inline ebib-utils.el:2396]: The empty line here is not necessary.

File: ebib-utils.el [hunk 5]
```diff
@@ -2376,25 +2392,19 @@ year can be extracted from DATE, return nil."
...
-            ("textbf" (setq arg (propertize arg 'face '(bold))))
-            ("textsc" (setq arg (upcase arg))))
-          (setq string (replace-match arg t t string)))))
-    ;; Replace as defined in `ebib-TeX-markup-replace-alist', and
-    ;; remove all {braces} (this takes care of nested braces too)
-    (replace-regexp-in-string
-     (rx-to-string (append '(or) (mapcar #'car ebib-TeX-markup-replace-alist)))
-     (lambda (match) (cdr (assoc match ebib-TeX-markup-replace-alist 'string=)))
-     string)))
+    (cl-loop for (ptn . rep) in ebib-TeX-markup-replace-alist
```
Comment (joostkremers) [inline ebib-utils.el:2403]: I'd prefer `pattern` and `replacement` here as variable names, TBH.

File: ebib-utils.el (hunk 3 above)
Comment (Hugo-Heagren) [reply ebib-utils.el:2399]: Done

File: ebib-utils.el (hunk 4 above)
Comment (Hugo-Heagren) [reply ebib-utils.el:2396]: Done

File: ebib-utils.el (hunk 5 above)
Comment (Hugo-Heagren) [reply ebib-utils.el:2403]: Can do.

File: ebib-utils.el (hunk 1 above)
Comment (Hugo-Heagren) [reply ebib-utils.el:399]: Good spot! No, there wasn't a reason. Because of the different method in the code they have to be strucutred slightly differently anyway, so I had to type them out by hand, and I think I just did it a different way. I'll revert to the originals.

Comment (Hugo-Heagren) [reply ebib-utils.el]: Done in latest force-push

**Comment by Hugo-Heagren** [comment]
> Even though they really are trivial, I think I'd still prefer them as properly defined functions, mainly because it's a user-configurable option.

Done.

> That would be an option. Whether it's worth it, I can't say right now. If it's just these three, then perhaps not. If the pattern is something that is more generally useful, then it might be. (Right now, though, I don't see how it would be more generally useful, TBH.)

Yeah this is my opinions too I think. Might be useful but for now not worth it.

> Emacs' Unicode support is quite good, so that's not an issue. The problem is with fonts. If the user doesn't have a font that supports a particular character (or has one that differs visually from their default font), then the display will be ugly and/or distorted (line height and alignment may be disrupted). So I wouldn't add too many characters by default, though a few common ones might be nice.

Ok that sounds good. One problem I had with this was decideig which symbols to include and which not to -- basically a judgement call about what's worth using. 

I think a good solution to this, and one which will definitely be compatible with many emacs installations, is to pull a list of conversions from the TeX input method. I don't have time right now to do this, but I've poked around a bit and I think it should be possible to generate an alist of valid TeX regexes, with corresponding unicode characters. What's great is that this is the list which is used in standard emacs, so will (I assume) mainly use widey-used, widely-supported` characters and forego others. If I can do it with a few lines of code and just ,@splice the list into the variable, it also means there can be a good number of common replacements defined by default, without having to list them manually in the source code (which is a pain).

What do you think?

> ... I assume adding a general regex for this case to the end of ebib-TeX-markup-replace-alist should be able to handle that.

Yes, added and rebased into the original conversion commit.

File: ebib-utils.el (hunk 2 above)
Comment (joostkremers) [reply ebib-utils.el:421]: Well, I do this in `writeroom-mode` here: https://github.com/joostkremers/writeroom-mode/blob/eac1da790f316f357ed76ed67fbb790d6a4d126a/writeroom-mode.el#L256, which I think was inspired by some Org mode customization options.

I'm not entirely sure it can be adapted to work in this case and what it would look like in the Customize buffer, but it may be worth looking into at some point. (Feel free to leave it to me, though. Customize is quite a beast, I'd understand if you wouldn't want to invest your time in it. :smile:)

Comment (joostkremers) [reply ebib-utils.el:421]: It then looks like this:

image

File: ebib-utils.el [hunk 6]
```diff
@@ -392,22 +392,33 @@ transformation function returns something that can be displayed."
...
 
-(defcustom ebib-TeX-markup-replace-alist '(("{" . "")
-                                           ("}" . "")
-                                           ("``" . "“")
+(defcustom ebib-TeX-markup-replace-alist '(("``" . "“")
 					   ("`" . "‘")
 					   ("''" . "”")
-					   ("'" . "’"))
+					   ("'" . "’")
+					   ("\\\\textit\\(?:\\[.*\\]\\)?{\\(.*?\\)}"
```
Comment (joostkremers) [inline ebib-utils.el:399]: AFAIK `textit` et. al. don't accept optional arguments (the LaTeX2e author manual doesn't mention them) so these regexes can be simplified to `"\\\\textit{\\(.*?\\)}"`.

File: ebib-utils.el [hunk 7]
```diff
@@ -392,22 +392,33 @@ transformation function returns something that can be displayed."
...
+					    . ebib--convert-tex-italics)
+					   ("\\\\emph\\(?:\\[.*\\]\\)?{\\(.*?\\)}"
+					    . ebib--convert-tex-italics)
+					   ("\\\\textbf\\(?:\\[.*\\]\\)?{\\(.*?\\)}"
+					    . ebib--convert-tex-bold)
+					   ("\\\\textsc\\(?:\\[.*\\]\\)?{\\(.*?\\)}"
+					    . ebib--convert-tex-small-caps)
+					   ("\\\\[[:alpha:]]+\\(?:\\[.*\\]\\)?{\\(.*?\\)}" . "\\1")
```
Comment (joostkremers) [inline ebib-utils.el:407]: I used `[a-zA-Z*]` as the character class because (La)TeX commands can only consist of ASCII letters and may optionally end in an asterisk. (Actually, technically I believe the asterisk is some kind of argument, but I don't understand the particulars). `[:alpha:]` misses the asterisk and includes accented characters and characters from non-Latin alphabets, which AFAIK are not allowed in command names.

The group capturing optional arguments should also be modified a bit, because commands can have more than one optional argument. The original regex has `\\(?:\\[.*?\\]\\)*`, which I think works better.

Or actually, come to think of it, your version should also work, because `\\[.*\\]` will match the *longest* sequence of characters starting with `[` and ending with `]`, which should capture all optional arguments in one go. It might even be faster. Ok, in that case, leave it as it is and consider it an improvement. :wink:

Boy, I wish I had invested time in writing some tests for these kinds of things... :smile:

**Comment by joostkremers** [comment]
> Ok that sounds good. One problem I had with this was decideig which symbols to include and which not to -- basically a judgement call about what's worth using.

Yup. And since I don't write math and usually use Unicode characters as much as possible, it's not something I can help with much.

[TeX input method]
> What's great is that this is the list which is used in standard emacs, so will (I assume) mainly use widey-used, widely-supported` characters and forego others.

Not so sure about that... `describe-input-method` for `TeX` gives me a list of 1553 characters (7 columns by 222 rows minus 1 because the last row contains only six characters). If you want to include all of those, it should be an option and it should probably be turned off by default, because I fear it would make filling the entry buffer unbearably slow.

So a hand-selected and curated list sounds more feasible, I think. Alternatively, one could check if the user has `prettify-symbols-alist` set up for TeX or LaTeX buffers and use that list instead. That will only work if the user has opened a `.tex` file before opening Ebib, however, so might also not be the ideal solution.

File: ebib-utils.el (hunk 7 above)
Comment (Hugo-Heagren) [reply ebib-utils.el]: Sorted. Thanks for explaining!

**Comment by Hugo-Heagren** [comment]
> Not so sure about that... describe-input-method for TeX gives me a list of 1553 characters

Fair point!!

> So a hand-selected and curated list sounds more feasible

I think this is the best way, especially as it doesn't rely on anything else. I think really all we need are symbols commonly used in writing title (so nothing maths-specific, probably nothing pictorial etc.). I went through the [comprehensive LaTeX symbols list](http://tug.ctan.org/info/symbols/comprehensive/symbols-a4.pdf) and included everything I thought fit the bill. For me at least, a big draw of setting this up was that I could render accents and diacritics properly, and that turned out to be surprisingly easy. So I've included what I think is relevant.

A note on the implementation: the way I've done accents is by replacing a latex command sequence like `\a{b}` with whatever was between the brackets (using a capture group), followed by a unicode combining character. This is simple and doesn't require writing or running extra functions. I don't know much about encoding though, so I don't know if storing the information in this way is ok, or robust on multiple platforms. Can you comment on this?

I also had to make `ebib-clean-TeX-markup` match case-sensitively, because LaTeX accent commands are case sensitive. Hope that's alright.

Other than that, I think this PR is good to go, so I'll mark it ready for review.

File: ebib-utils.el [hunk 8]
```diff
@@ -2375,26 +2440,33 @@ year can be extracted from DATE, return nil."
...
+  "Return STRING without TeX markup.
+Any substring matching the car of a cell in
+`ebib-TeX-markup-replace-alist' is replaced with the
+correspnoding cdr (if the cdr is a string), or with the result of
+calling the cdr on the match (if it is a function). This is done
+with `replace-regexp-in-string', which see for details."
+  (let ((case-fold-search nil))
+    (save-match-data
+      (cl-loop for (pattern . replacement) in ebib-TeX-markup-replace-alist
+	       if (string-match pattern string)
```
Comment (joostkremers) [inline ebib-utils.el:2465]: I don't think the `if` here is needed, since `replace-regexp-in-string` just returns the string unmodified if no matches are found.

**Comment by joostkremers** [comment]
> A note on the implementation: the way I've done accents is by replacing a latex command sequence like `\a{b}` with whatever was between the brackets (using a capture group), followed by a unicode combining character. This is simple and doesn't require writing or running extra functions. I don't know much about encoding though, so I don't know if storing the information in this way is ok, or robust on multiple platforms. Can you comment on this?

Well, we'll just have to assume that using Unicode is OK (which I think it is), and that whatever encoding the system Emacs is running on uses, Emacs will Do The Right Thing™ when retrieving the source files from Melpa (which I also think it does).

In the end, the problem will mostly be the font that is being used, but that's why it should be possible to modify / disable these substitutions, which is the case.

> I also had to make `ebib-clean-TeX-markup` match case-sensitively, because LaTeX accent commands are case sensitive. Hope that's alright.

I don't see why not.

> Other than that, I think this PR is good to go, so I'll mark it ready for review.

There one complicating factor that we haven't considered yet. There are several different ways to write accents and non-ASCII letters in LaTeX. You can write `\^{a}`, but you can just as easily write `\^a`. And non-ASCII letters such as "å" can be written as `\aa ` (note the space) or `\aa{}`.

I guess such cases can be handled with regexps, so they shouldn't be too difficult to handle.

File: ebib-utils.el (hunk 8 above)
Comment (Hugo-Heagren) [reply ebib-utils.el]: Fixed

File: ebib-utils.el (hunk 6 above)
Comment (Hugo-Heagren) [reply ebib-utils.el]: Fixed in most recent push.

**Comment by Hugo-Heagren** [comment]
> There one complicating factor that we haven't considered yet. There are several different ways to write accents and non-ASCII letters in LaTeX. You can write `\^{a}`, but you can just as easily write `\^a`. And non-ASCII letters such as "å" can be written as `\aa ` (note the space) or `\aa{}`.

Good spot!

> I guess such cases can be handled with regexps, so they shouldn't be too difficult to handle.

Well, they're relatively straightforward, but the resulting regexps get very long. For example:

```elisp
 "\\\\\"\\(?:{\\(?1:[[:alpha:]]\\)}\\|\\(?1:[[:alpha:]]\\)\\(?:{}\\)?\\)"
```

This is unweildy in the source and difficult to understand or maintain. So I wrote some convenience functions with `rx` for generating pairs of regexp and replacement. This generation is done when the variable is initialised, so shouldn't slow general _use_ of ebib down considerably (I suppose it shouldn't make much of a difference anyway unless a user specificies a *lot* of extra substitutions). I had to put the functions directly before the variable definition for them to be available in it.

**Comment by joostkremers** [comment]
I found a solution to the question of Unicode characters in the source file: Elisp has several notations for Unicode characters that can be used, I chose the most descriptive one: `\N{COMBINING DIAERESIS}`. I pushed a commit to your PR to use this notation, but it isn't finished yet. (So far, it only deals with combining characters, i.e., the accents. I'd like to replace the other non-ASCII characters as well.)

I also pushed a few doc string fixes and added dotless i and j (though I forgot to use the `\N{...}` notation for them...)

I'm also working on a test suite for `ebib-clean-TeX-markup`, which is already revealing a couple of problems. :confused:

**Comment by joostkremers** [comment]
I just push a first version of the test set. One thing that currently stands out: commands for diacritics should be treated differently depending on whether they consist of a letter or a non-letter character. Basically, with non-letter commands, a space is optional, with letter-commands, it's obligatory. So `\'a` is fine, and so is `\' a`, but `\ua` is not. You need to write `\u a`.

This case needs to be fixed before I can merge the PR. I haven't gone through the other tests yet, but they're mostly not show stoppers, I think.

**Comment by Hugo-Heagren** [comment]
> I found a solution to the question of Unicode characters in the source file: Elisp has several notations for Unicode characters that can be used, I chose the most descriptive one: `\N{COMBINING DIAERESIS}`.

Nice!

> I also pushed a few doc string fixes and added dotless i and j (though I forgot to use the `\N{...}` notation for them...)

Also nice!

> I'm also working on a test suite for `ebib-clean-TeX-markup`, which is already revealing a couple of problems.

This is great. I don't have much (any) experience with programmatic testing so I don't really know how I could help, but I'll take a look when I can and try to sort out the problems.

**Comment by Hugo-Heagren** [comment]
[problems with letter/nonletter commands]

Great, I'll take a look at this later today. I don't think it should be too difficult to fix though, because a similar differentiation is already made between the accent and non-accent commands (the latter don't take args and so don't have the same behaviour), so I can probably just adapt that.

**Comment by joostkremers** [comment]
> > I'm also working on a test suite for `ebib-clean-TeX-markup`, which is already revealing a couple of problems.
> 
> This is great. I don't have much (any) experience with programmatic testing so I don't really know how I could help, but I'll take a look when I can and try to sort out the problems.

My experience is also very limited, but the basics aren't very difficult, I think. Emacs' test tool (ERT) is described in the manual, relevant would be the section on running tests.

If you want to run tests it inside an Emacs session, you should start Ebib (you'll need to load the code you're testing, of course :smile:) and then open the file with tests (`test/ebib-test.el`). Then use `C-M-x` to evaluate the tests you want to run. Evaluating a test just loads it, it doesn't run it yet. To run a test, do `M-x ert` and select the test you want to run.

You could also evaluate the entire file `ebib-test.el`, but you'll need a 3rd-party package `with-simulated-input` for that, so you may want to avoid it. (I should probably split the test suite into interactive and non-interactive tests; `with-simulated-input` is only needed for the former.)

**Comment by Hugo-Heagren** [comment]
I've started work on fixing the failing cases. I've done the dashes so far. Importantly, for the sake of a clean commit history, I'm rebasing my new changes into older commits, and force-pushing to my own branch. This should work fine, so long as we both make sure we always have the most up to date version of the branch before pushing our work.

**Comment by joostkremers** [comment]
> I've started work on fixing the failing cases. I've done the dashes so far. 

:+1:

> Importantly, for the sake of a clean commit history, I'm rebasing my new changes into older commits, and force-pushing to my own branch. 

I only speak basic git, so I'm not entirely sure what that means... :confused: Do I understand correctly that you're basically replacing existing commits with new&improved versions? Or inserting new commits in between existing ones?

> This should work fine, so long as we both make sure we always have the most up to date version of the branch before pushing our work.

OK, should I decide to make any more changes, I'll give you a heads-up and create a fresh topic branch.

**Comment by Hugo-Heagren** [comment]
> I only speak basic git, so I'm not entirely sure what that means... confused Do I understand correctly that you're basically replacing existing commits with new&improved versions?

Exactly, yes. Sorry, I'm never quite sure how to pitch this sort of thing with other coders -- I consider myself a rank amateur and assume everyone knows more than me. Here's a more complete explanation:
- I pull the most recent version of the project from [`Hugo-Heagren:better-TeX-markup-replacements`](https://github.com/Hugo-Heagren/ebib/tree/better-TeX-markup-replacements). I say most recent because for example, you might have worked on it and added commits.
- I do some work, which edits an old commit, replacing it with a new, improved version. The content will be different, but often the comitt message will be the same. This changes the commit's hash, *and* the hashes of *all subsequent commits on that branch*. (this was my comment about rebasing. Technically the git command used to include new changes in old commit is `rebase`, though how I use it isn't relevant)
- So, when I push my new work to github (pushing to [`Hugo-Heagren:better-TeX-markup-replacements`](https://github.com/Hugo-Heagren/ebib/tree/better-TeX-markup-replacements)) I have to use `--force`. This replaces all the commits on github with the old hashes with those with the new hashes (though the commit messages, and the content of the commits I didn't edit, will be the same).

*So*, if I do all that and you want to contribute something new to the branch, you need to make sure you're working with the most recent version -- i.e. the version with my edit commit and the updated commit hashes, not the old commit with the old hashes (even though the messages and contents will look very similar). Otherwise when you push, github won't recognise that the commits in your branch line up with the commits in the one I've just pushed (because the hashes will be different).

Hope that helps!

> OK, should I decide to make any more changes, I'll give you a heads-up and create a fresh topic branch.

That'll work, though if you're just making new commits on top of my work you should be fine working on the same branch, so long as you make sure you have the most up-to-date version before you push.

**Comment by joostkremers** [comment]
> I consider myself a rank amateur and assume everyone knows more than me. 

Heh, I could have said that about myself. :smile:

> Here's a more complete explanation:
> [snip]
> Hope that helps!

Thanks, yes, that gives me a better picture.

This way, I won't be able to see exactly what you changed, though, will I? Or is there some way I can see a diff between an old commit and the new one that replaces it? Not sure if I'd ever want to, just wondering.

BTW, if you're working with the test suite, if one fails, please also check if the test itself is correct. I tried to be careful when I created them, but I'm only human...

**Comment by Hugo-Heagren** [comment]
> This way, I won't be able to see exactly what you changed, though, will I? Or is there some way I can see a diff between an old commit and the new one that replaces it? Not sure if I'd ever want to, just wondering.

That's right. I think generally with small stuff like that's ok. The other way people sometimes do something similar is to *always* make new commits until the PR is definitely ready to go, then go back and 'tidy up' the history with git's 'squash' feature. This has the same effect, though it means there are a lot of commits floating around in the meantime. I could do it that way though if it would be clearer?

> BTW, if you're working with the test suite, if one fails, please also check if the test itself is correct. I tried to be careful when I created them, but I'm only human...

Right, so I've fixed all the tests except `ebib-clean-TeX-markup-escapable-characters` and `ebib-clean-TeX-markup-other-commands`.

`ebib-clean-TeX-markup-escapable-characters` I don't think is quite right -- it says that multiple whitespace should be retained in certain situations, but (without { } braces {     }  like {    } this), LaTeX doesn't do this. I tried putting the tests's string into a document and in the compiled pdf each run of whitespace had been collapsed to a single space. (that's reflected in [this new rule](https://github.com/joostkremers/ebib/pull/248/commits/a728e0e36cadd4b4757f128fa0fd7a230f54e1d1)). 

I think there's a slight inconsistency in `ebib-clean-TeX-markup-other-commands`. Unknown command with optional arguments are meant to printed *without* a leading backslash (so `\foo[bar][baz]` => `foo`), but commands with no optional arguments are printed *with* a leading backslash (so `\LaTeX{}` => `\LaTeX`). Is this intentional?

Other than those, I've fixed all the tests, and I've just pushed the new commits.

**Comment by joostkremers** [comment]
> That's right. I think generally with small stuff like that's ok. The other way people sometimes do something similar is to _always_ make new commits until the PR is definitely ready to go, then go back and 'tidy up' the history with git's 'squash' feature. This has the same effect, though it means there are a lot of commits floating around in the meantime. I could do it that way though if it would be clearer?

Nah, if the commits were large it would be a different issue, but they're not, so it's fine.

> `ebib-clean-TeX-markup-escapable-characters` I don't think is quite right -- it says that multiple whitespace should be retained in certain situations, but (without { } braces { } like { } this), LaTeX doesn't do this.

Good point. Handling of spaces should be consistent, so they should always be collapsed into a single space, I guess.

> I tried putting the tests's string into a document and in the compiled pdf each run of whitespace had been collapsed to a single space. (that's reflected in [this new rule](https://github.com/joostkremers/ebib/pull/248/commits/a728e0e36cadd4b4757f128fa0fd7a230f54e1d1)).

If I click on that link, Github tells me the commit can't be found. (It says this can happen when commits are force-pushed...)

> I think there's a slight inconsistency in `ebib-clean-TeX-markup-other-commands`. Unknown command with optional arguments are meant to printed _without_ a leading backslash (so `\foo[bar][baz]` => `foo`), but commands with no optional arguments are printed _with_ a leading backslash (so `\LaTeX{}` => `\LaTeX`). Is this intentional?

Nope, that's one of those human failings I hinted at. :smile: I'd suggest to keep the backslash in both cases. (It might even make more sense to keep the optional arguments as well, I don't know. I don't have any actual examples for such cases...)

> Other than those, I've fixed all the tests, and I've just pushed the new commits.

Great, thanks! I'll try and check out the code tomorrow or the day after.

**Comment by joostkremers** [comment]
I pushed a few commits, some of which are purely cosmetic, some are more substantial.

I fixed the test for escapable characters (collapsed white space in the expected outputs) and for commands with optional arguments and empty braces (added backslash).

There was also one test that was still failing, specifically the one that tested commands with empty braces, i.e., `\LaTeX{}`. To make that work, I added another clause to `ebib-TeX-markup-replace-alist`: there's now a clause that deals with commands with obligatory arguments and one specifically for commands with empty braces.

At this point, all the tests succeed, so I'd say this can now be merged. :slightly_smiling_face:. 

One problem I still see is that `ebib-TeX-markup-replace-alist` is now so complex that it's hardly suitable for a user option in the current form. I have an idea on how to improve it, but I don't want to let that delay the merge.

I also updated the doc strings for `ebib--build-TeX-accent-command-regexp` and `ebib-build-TeX-command-regexp`. There were some errors in them about which kinds of commands are matched, and then I ended up rephrasing them more substantially. Hope you don't mind...

Let me know if there's anything that you still want to add or update. If so, I'll hold off merging until you're done. If not, I'll go ahead and do the merge.

**Comment by Hugo-Heagren** [comment]
> I fixed the test for escapable characters (collapsed white space in
> the expected outputs) and for commands with optional arguments and
> empty braces (added backslash).

Nice, that sounds like sensible behaviour.

> There was also one test that was still failing, specifically the one
> that tested commands with empty braces, i.e., `\LaTeX{}`. To make
> that work, I added another clause to
> `ebib-TeX-markup-replace-alist`: there's now a clause that deals
> with commands with obligatory arguments and one specifically for
> commands with empty braces.

Also sensible.


> I also updated the doc strings for
> `ebib--build-TeX-accent-command-regexp` and
> `ebib-build-TeX-command-regexp`. There were some errors in them about
> which kinds of commands are matched, and then I ended up rephrasing
> them more substantially. Hope you don't mind...

Yeah I'm sometimes a bit forgetful about updating documentation, especially when I'm going back and making changes to old stuff. Sorry about that... Thanks for fixing it!

> One problem I still see is that `ebib-TeX-markup-replace-alist` is
> now so complex that it's hardly suitable for a user option in the
> current form.

That's true. Though the variable itself isn't two complex right? It's just an alist of regexps and replacements---anyone comfortable with regexp (which is admittedly a bit complex in elisp) could add a couple of commands right?

> I have an idea on how to improve it, but I don't want to let that
> delay the merge.

I'd be interested to hear! Though I agree, it would be good to merge the functionality now and make it more user-friendly later.

> At this point, all the tests succeed, so I'd say this can now be
> merged. :slightly_smiling_face:.

> Let me know if there's anything that you still want to add or update.
> If so, I'll hold off merging until you're done. If not, I'll go ahead
> and do the merge.

Sounds good to me! There's nothing more I want to add, so go ahead and merge! 😀

**Comment by joostkremers** [comment]
> That's true. Though the variable itself isn't two complex right? It's just an alist of regexps and replacements---anyone comfortable with regexp (which is admittedly a bit complex in elisp) could add a couple of commands right?

You can easily use `add-to-list` in your init file to add something, yes. Most people comfortable with regexes may do it that way. If you look at the variable in a Customise buffer, though, you'll get a long and daunting list of regexes and no clue as to what they do.

> I'd be interested to hear! Though I agree, it would be good to merge the functionality now and make it more user-friendly later.

What I would like (but honestly, yesterday I thought it would be fairly easy to implement, right now I'm not so sure anymore...) is a list of check boxes for the categories that you distinguish with comments in `ebib-TeX-markup-replace-alist`. I thought this could be done with the `:set` keyword of `defcustom`, but now I'm not so sure anymore.

I'll need to think about it a bit.

> Sounds good to me! There's nothing more I want to add, so go ahead and merge! 

Done! Thanks for your work on this!
